File: workspace/2023/2023-02-07/parameterized-testing.python/parameterized.py
import inspect
import unittest


def parameterized(tuples):
    def new_wrapper(method, tuple, i):
        wrapper = lambda self: method(self, *tuple)
        wrapper.__name__ = f"{method.__name__}_{i}"
        return wrapper

    def decorator(method):
        for i, tuple in enumerate(tuples):
            wrapper = new_wrapper(method, tuple, i)
            f_locals = inspect.currentframe().f_back.f_locals
            f_locals[wrapper.__name__] = wrapper

    return decorator


class TestCase(unittest.TestCase):

    @parameterized([
        # (expected, a, b)
        ( 3, 1, 2),
        ( 7, 3, 4),
        (11, 5, 6),
        ( 0, 7, 8),
    ])
    def test(self, expected, a, b):
        self.assertEqual(expected, a + b)


if __name__ == "__main__":
    unittest.main()

File: workspace/2023/2023-04-01/parameterized-testing.ruby/parameterized.rb

def parameterized(name, arguments_expected_map)
    arguments_expected_map.each_with_index do |(arguments, expected), index|
        define_method("test_#{name}_#{index + 1}") do
            actual = method(name).call(*arguments)
            assert_equal(expected, actual)
        end
    end
end

